modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -105,11 +104,50 @@ object Sink {
     capDelay[F](config.maxBackoff, fullJitter[F](config.minBackoff))
       .join(limitRetries(config.maxRetries))
 
+  private def writeToKinesis[F[_]: ContextShift: Parallel: Sync: Timer](
+    blocker: Blocker,
+    config: Output.Kinesis,
+    kinesis: AmazonKinesis,
+    records: List[AttributedData[Array[Byte]]],
+    retryPolicy: RetryPolicy[F]
+  ): F[Unit] =
+    group(toKinesisRecords(records), config.recordLimit, putRecordsSizeLimit, getRecordSize)
+      .parTraverse_(g => writeToKinesis(blocker, config, kinesis, g, retryPolicy))
+
+  /**  This function takes a list of records and splits it into several lists,
+    *  where each list is as big as possible with respecting the record limit and
+    *  the size limit.
+    */
+  private[kinesis] def group[A](
+    records: List[A],
+    recordLimit: Int,
+    sizeLimit: Int,
+    getRecordSize: A => Int
+  ): List[List[A]] = {
+    case class Batch(size: Int, count: Int, records: List[A])
+
+    records.foldLeft(List(Batch(0, 0, Nil))){ case (acc, record) =>
```

Comment:
Would it be safer to do `foldLeft(List.empty[Batch]) {` in case `records` is an empty list?

Reply:
I doubt that `records` can be empty but good point, better to be safe than sorry!

Reply:
> I doubt that records can be empty

Actually, I think it can be:

1. If this is the good sink, and all the records in the batch were bad
2. If this is the bad sink, and all the records in the batch were good

But your latest version looks safe.